content/post/review-building-static-analysis-tools.md

```diff
@@ -0,0 +1,249 @@
+
+For this kind of tools, different integration points in a software development process of an organization exist, where results could be presented to the developers. They are discussed in the paper in the chronological order of BugBot team’s attempts of integrating an existing OSS static analysis tool for Java — [FindBugs](https://github.com/findbugsproject/findbugs) into the Google’s developer workflow.
+
+## History of integrating FindBugs at Google
+
+### 0. IDE/Editor
+
+Using editor as a place to show results of analysis was discarded from scratch.
+
+Cons:
+
+* many different editors/IDEs to support, as it’s before the build time
+
+* wrong time — distracting developer who may not be open to the feedback
+
+Interesting enough, [source{d}](http://sourced.tech/) as well experimented with this approach using some neural-network-based models [https://github.com/src-d/code-completion](https://github.com/src-d/code-completion) a while ago and came to the very same conclusions.
+ 
+Personally, I belive there is a huge room for improvement of code reading/writing experience though, enhancing or better augmenting developer’s abilities to both [navigate](http://kythe.io/), [understand](http://moosetechnology.org/) and [create](http://witheve.com/) programs. But indeed, existing static analysis tools which target defect detection are hardly the right tools for any of these jobs.
+
+### 2006 Bug Dashboard
+
+Next attempt included results of analysis aggregated and served from a separate web dashboard.
+
+Cons:
+
+* outside the developers’ usual workflow
+
+* distinguishing between new and existing static-analysis issues was distracting
+
+This resonates with my personal experience — in 2007, as a Build Engineer on big Java project at the first place of job, a consultancy outlet, I have experience creating internal **“Personalized Quality Dashboard”** service.
+
+It was a bunch of static HTML files, that later became a database for a web application, produced as a part of nightly build using Maven that were copied and served from a well known URL inside the company. HTMLs would contain tables of potential software defects, obtained by running multiple existing static analysis tools — [FindBugs](https://github.com/findbugsproject/findbugs), [PMD](https://pmd.github.io/) and [CheckStyle](https://github.com/checkstyle/checkstyle). Each defect was attributed to the latest change in the codebase using “git blame” and “assigned” to a particular engineer who introduced the change.
+
+It’s internal adoption numbers, although driven top-down by the management decision, were *very low* and aligned with the paper — needless to say that not many engineers were motivated enough to go to a separate [http://code-quality.company.com](http://code-quality.company.com) every day, only to find that they are #N by the amount of bugs introduced to the codebase.
```

Comment:
It’s internal adoption numbers?..

Finally understood that you mean the Personalized Quality Dashboard. I suggest to avoid "it" here - very confusing.

Reply:
go to **the** separate

Reply:
PAIN!!! I feel sorry for your first job Alex.

Reply:
Refactored